src/metrics/include/metrics/MetricsReporter.hpp

```diff
@@ -22,108 +22,106 @@
+        : _registry{std::addressof(registry)} {}
 
     /** @return how many distinct gauges were registered */
-    auto getGaugeCount(v1::Permission perm = {}) const {
+    auto getGaugeCount() const {
+        v1::Permission perm;
         auto& x = _registry->getGauges(perm);
         return std::distance(x.begin(), x.end());
     }
 
     /** @return how many gauge data-points were recorded */
-    long getGaugePointsCount(v1::Permission perm = {}) const {
+    long getGaugePointsCount() const {
+        v1::Permission perm;
         return dataPointsCount(_registry->getGauges(perm), perm);
     }
 
     /** @return how many distinct timers were registered */
-    auto getTimerCount(v1::Permission perm = {}) const {
+    auto getTimerCount() const {
+        v1::Permission perm;
         auto& x = _registry->getTimers(perm);
         return std::distance(x.begin(), x.end());
     }
 
     /** @return how many timer data-points were recorded */
-    long getTimerPointsCount(v1::Permission perm = {}) const {
+    long getTimerPointsCount() const {
+        v1::Permission perm;
         return dataPointsCount(_registry->getTimers(perm), perm);
     }
 
     /** @return how many counters were registered */
-    auto getCounterCount(v1::Permission perm = {}) const {
+    auto getCounterCount() const {
+        v1::Permission perm;
         auto& x = _registry->getCounters(perm);
         return std::distance(x.begin(), x.end());
     }
 
     /** @return how many counter data-points were recorded */
-    long getCounterPointsCount(v1::Permission perm = {}) const {
+    long getCounterPointsCount() const {
+        v1::Permission perm;
         return dataPointsCount(_registry->getCounters(perm), perm);
     }
 
+private:
+    struct SystemClockSource {
+        using clock = std::chrono::system_clock;
+
+        static std::chrono::time_point<clock> now() {
+            return clock::now();
+        }
+    };
+
+public:
     /**
      * @param out print a human-readable listing of all
      *            data-points to this ostream.
-     * @param metricsFormat the format to use. Either "csv" or "sys-perf".
-     * @param perm passkey permission (must be friends with metrics::Registry)
+     * @param metricsFormat the format to use. Must be "csv".
      */
-    void report(std::ostream& out,
-                const std::string& metricsFormat,
-                v1::Permission perm = {}) const {
+    template <typename ReporterClockSource = SystemClockSource>
+    void report(std::ostream& out, const std::string& metricsFormat) const {
+        v1::Permission perm;
+
         // should these values come from the registry, and should they be recorded at
         // time of registry-creation?
-        auto systemTime = std::chrono::system_clock::now().time_since_epoch().count();
-        auto metricsTime = _registry->now(perm).time_since_epoch().count();
+        auto systemTime = std::chrono::duration_cast<std::chrono::nanoseconds>(
```

Comment:
It looks like we were calling `.count()` on the `std::chrono::duration` returned by the `.time_since_epoch()` without knowing what units were being used for the system clock. It turns out to be different between Amazon Linux 2 and Arch Linux (nanoseconds) and RHEL 7.0, Ubuntu 18.04, and macOS (microseconds).

Reply:
The server codebase defines [a `durationCount()` helper function](https://github.com/mongodb/mongo/blob/r4.1.8/src/mongo/util/duration.h#L125-L141) to make it so the number of ticks is logged according to the intended units. I added a `nanosecondsCount()` helper function since we don't need the extra parameterization.

> consider making the unit a typedef here so refactoring couldn't cause the units of metricsTime and systemTime to change independently.

I'd rather we say "nanoseconds" so it is more obvious to future readers what the units in the output are without needing to look at another typedef. (I also plan to add a header for the "Clocks" section that says "nanoseconds" in the "csv" file too.)